File: src/Lox.cpp
#include "Scanner.hpp"

#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sysexits.h>

namespace fs = std::filesystem;

class Lox {
public:
  bool hadError = false;

  Lox() = default;

  void runFile(const std::string &path) const {
    auto file = Lox::openFileOrExit(path);
    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string content = buffer.str();
    std::cout << "File contents:\n---\n" << content << "\n---\n";
    run(content);

    if (hadError) {
      exit(EX_DATAERR);
    }
  }

  void runPrompt() {
    while (true) {
      std::string readLine;
      std::getline(std::cin, readLine);
      if (readLine.empty()) {
        break;
      }
      run(readLine);
    }
  }

private:
  static void run(std::string_view source) {
    Scanner scanner(source);
    auto tokens = scanner.scanTokens();
    for (const auto &token : tokens) {
      std::cout << token.content << "\n";
    }
  }

  void report(int line, std::string &where, std::string &message) {
    std::cout << std::format("[line {}] Error{}: {}\n", line, where, message);
    hadError = true;
  }

  void error(int line, std::string &message) {
    auto where = std::string("");
    report(line, where, message);
  }

  static std::ifstream openFileOrExit(const std::string &path) {
    fs::path filePath(path);
    std::ifstream file(filePath);

    if (!file.is_open()) {
      std::cerr << std::format("Error opening file: {}\n", path);
      exit(1);
    }

    return file;
  }
};

File: src/main/kotlin/com/craftinginterpreters/lox/Lox.kt
package com.craftinginterpreters.lox

import java.io.File
import kotlin.system.exitProcess
import java.util.Scanner as InputScanner

object Lox {
    var hadError = false

    fun runFile(path: String) {
        val file = File(path)
        if (!file.exists()) {
            System.err.println("Error opening file: $path")
            exitProcess(1)
        }
        val content = file.readText()
        println("File contents:\n---\n$content\n---")
        run(content)

        if (hadError) {
            exitProcess(65) // EX_DATAERR
        }
    }

    fun runPrompt() {
        val scanner = InputScanner(System.`in`)
        while (true) {
            print("> ")
            if (!scanner.hasNextLine()) break
            val line = scanner.nextLine()
            if (line.isEmpty()) break
            run(line)
            hadError = false
        }
    }

    private fun run(source: String) {
        val scanner = Scanner(source)
        val tokens = scanner.scanTokens()
        for (token in tokens) {
            println(token.lexeme)
        }
    }

    fun error(
        line: Int,
        message: String,
    ) {
        report(line, "", message)
    }

    private fun report(
        line: Int,
        where: String,
        message: String,
    ) {
        System.err.println("[line $line] Error$where: $message")
        hadError = true
    }
}

fun main(args: Array<String>) {
    try {
        if (args.size > 1) {
            println("Usage: lox_interpreter [script]")
            exitProcess(64) // EX_USAGE
        } else if (args.size == 1) {
            Lox.runFile(args[0])
        } else {
            Lox.runPrompt()
        }
    } catch (e: Exception) {
        System.err.println("Fatal error: ${e.message}")
        exitProcess(1)
    }
}
